services/crawler/crawler.go

```diff
@@ -0,0 +1,104 @@
+package crawler
+
+import (
+	"strings"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/beego/beego/v2/core/logs"
+)
+
+const googleSearchBaseUrl = "https://www.google.com/search"
+
+var doc *goquery.Document
+var htmlResponse string
+var selectorMapping = map[string]string{
+	"topAdWordAdvertisersCount":   "#tads .uEierd",
+	"topAdWordAdvertisersUrls":    "#tads .Krnil",
+	"totalAdWordAdvertisersCount": ".Krnil",
+	"resultsCount":                "#rso .yuRUbf",
+	"resultsUrls":                 "#rso .yuRUbf > a",
+	"totalLinksCount":             "a",
+}
+
+type CrawlData struct {
+	TopAdWordAdvertisersCount   int
+	TopAdWordAdvertisersUrls    []string
+	TotalAdWordAdvertisersCount int
+	ResultsCount                int
+	ResultsUrls                 []string
+	TotalLinksCount             int
+	Html                        string
+}
+
+// Call this function with your search key, it will return the necessary crawled data
+// Ex: Crawl("Buy laptop")
+func Crawl(searchString string) (data *CrawlData, err error) {
+	searchUrl, err := BuildSearchUrl(searchString)
+	if err != nil {
+		return &CrawlData{}, err
+	}
+
+	response, err := GetRequest(searchUrl)
+	if err != nil {
+		return &CrawlData{}, err
+	}
+
+	htmlResponse = string(response)
+	doc, err = goquery.NewDocumentFromReader(strings.NewReader(htmlResponse))
+	if err != nil {
+		logs.Error("Initializing Goquery document failed: ", err)
+	}
+
+	data = parseCrawledData()
+
+	return data, err
```

Comment:
Is this the success case? If yes I suggest changing this to `return data, nil`. In this case, `err` also `nil` but this makes me confused when reading the code.